### PR TITLE
Feature/csharp api 4

### PR DESCRIPTION
**Pull request by KerstinKeller**
**Pull request type**
- [x] Feature

This PR introduces 4 commits which should be rebased (not squashed) which each introduces new C# Features:

1. Build C# as standalone project:
   With this commit, the C# project can be build as a standalone project, against an eCAL SDK (tested against internal conan build)
   Also, .proto files are compiled to C# files, e.g. they are not committed with the source code.
   
 2. Add additional API functionality for performance
    Add some APIs to interface to C# via Bytes as opposed to String (e.g. C# Strings are not just byte containers)

3. Add reflection  (descriptor) information to C# Publishers
   So finally C# Publishers / Subscribers topics can be inspected in eCAL Mon

4. Wrap Service / Client interface to C#
    Including samples (minimal). Unfortunately, protobuf is not supported as protoc ignores services when generating C# code.

### CODE REVIEW

File: lang/csharp/Continental/eCAL/Core/ecal_clr.cpp [hunk 1]
```diff
@@ -41,7 +41,21 @@ std::string StringToStlString(System::String^ s_)
...
+  GCHandle handle = GCHandle::Alloc(array_, GCHandleType::Pinned);
+  size_t len = array_->Length;
+  std::string ret((const char*)(void*)handle.AddrOfPinnedObject(), len);
+  handle.Free();
+  return(ret);
+}
 
+array<Byte>^ StlStringToByteArray(const std::string& string_)
+{
+  array<Byte>^ array_ = gcnew array<Byte>(string_.size());
```
[the remark this answers is not in the document]
Comment (FlorianReimold) [reply lang/csharp/Continental/eCAL/Core/ecal_clr.cpp:55]: False positive

File: lang/csharp/Continental/eCAL/Core/ecal_clr.cpp [hunk 2]
```diff
@@ -41,7 +41,21 @@ std::string StringToStlString(System::String^ s_)
   return(s);
 }
 
+std::string ByteArrayToStlString(array<Byte>^ array_)
+{
+  GCHandle handle = GCHandle::Alloc(array_, GCHandleType::Pinned);
+  size_t len = array_->Length;
```
[the remark this answers is not in the document]
Comment (FlorianReimold) [reply lang/csharp/Continental/eCAL/Core/ecal_clr.cpp:47]: False positive

File: lang/csharp/Continental/eCAL/Core/ecal_clr.cpp [hunk 3]
```diff
@@ -41,7 +41,21 @@ std::string StringToStlString(System::String^ s_)
   return(s);
 }
 
+std::string ByteArrayToStlString(array<Byte>^ array_)
+{
+  GCHandle handle = GCHandle::Alloc(array_, GCHandleType::Pinned);
```
[the remark this answers is not in the document]
Comment (FlorianReimold) [reply lang/csharp/Continental/eCAL/Core/ecal_clr.cpp:46]: False positive

File: lang/csharp/Continental/eCAL/Core/ecal_clr.cpp [hunk 4]
```diff
@@ -41,7 +41,21 @@ std::string StringToStlString(System::String^ s_)
...
+  size_t len = array_->Length;
+  std::string ret((const char*)(void*)handle.AddrOfPinnedObject(), len);
+  handle.Free();
+  return(ret);
+}
 
+array<Byte>^ StlStringToByteArray(const std::string& string_)
```
[the remark this answers is not in the document]
Comment (FlorianReimold) [reply lang/csharp/Continental/eCAL/Core/ecal_clr.cpp:53]: False positive

File: samples/csharp/person/protobuf/protobuf_csharp_helper.cmake [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+  # let's see if this will work for CS  
+  #set_source_files_properties(${INPUT_PROTO_FILES} PROPERTIES HEADER_FILE_ONLY TRUE)
+  #source_group(protobuf\\proto FILES
+  #  ${INPUT_PROTO_FILES} 
+  #)
+endfunction()
+
+
+function(PROTOBUF_TARGET_CSHARP TARGET_NAME PROTO_ROOT)
+  message(STATUS "Hello!!!")
```
Comment (hannemn) [inline samples/csharp/person/protobuf/protobuf_csharp_helper.cmake:96]: Unintended debug message :-)

File: lang/csharp/Continental/eCAL/Core/ecal_clr.cpp [hunk 6]
```diff
@@ -309,8 +388,130 @@ void Subscriber::OnReceive(const char* topic_name_, const ::eCAL::SReceiveCallba
...
+/////////////////////////////////////////////////////////////////////////////
+// ServiceClient
+/////////////////////////////////////////////////////////////////////////////
+ServiceClient::ServiceClient() : m_client(new ::eCAL::CServiceClient())
+{
+}
+
+ServiceClient::ServiceClient(System::String^ service_name_)
+{
+    m_client = new ::eCAL::CServiceClient(StringToStlString(service_name_));
```
Comment (hannemn) [inline lang/csharp/Continental/eCAL/Core/ecal_clr.cpp:474]: Replace with unique_ptr if possible.

File: lang/csharp/Continental/eCAL/Core/CMakeLists.txt [hunk 7]
```diff
@@ -18,6 +18,8 @@
 
 project(Continental.eCAL.Core)
 
+find_package(eCAL REQUIRED)
```
Comment (hannemn) [inline lang/csharp/Continental/eCAL/Core/CMakeLists.txt:21]: Only the eCAL::core target is required for building the binding.

File: lang/csharp/Continental/eCAL/Core/ecal_clr.cpp [hunk 8]
```diff
@@ -142,6 +156,11 @@ Publisher::Publisher(System::String^ topic_name_, System::String^ topic_type_, S
   m_pub = new ::eCAL::CPublisher(StringToStlString(topic_name_), StringToStlString(topic_type_), StringToStlString(topic_desc_));
 }
 
+Publisher::Publisher(System::String^ topic_name_, System::String^ topic_type_, array<Byte>^ topic_desc_)
+{
+  m_pub = new ::eCAL::CPublisher(StringToStlString(topic_name_), StringToStlString(topic_type_), ByteArrayToStlString(topic_desc_));
```
Comment (hannemn) [inline lang/csharp/Continental/eCAL/Core/ecal_clr.cpp:161]: Use RAII pattern if possible.

File: lang/csharp/Continental/eCAL/Core/CMakeLists.txt (hunk 7 above)
Comment (KerstinKeller) [reply lang/csharp/Continental/eCAL/Core/CMakeLists.txt:21]: Yes, but we don't really handle components.... so what do you suggest?

**Comment by KerstinKeller** [comment]
I will make four subsequent PRs so we can get things merged step by step.

File: lang/csharp/Continental/eCAL/Core/ecal_clr.cpp [hunk 9]
```diff
@@ -309,8 +388,130 @@ void Subscriber::OnReceive(const char* topic_name_, const ::eCAL::SReceiveCallba
...
+/////////////////////////////////////////////////////////////////////////////
+// ServiceServer
+/////////////////////////////////////////////////////////////////////////////
+ServiceServer::ServiceServer() : m_serv(new ::eCAL::CServiceServer())
+{
+}
+
+ServiceServer::ServiceServer(System::String^ server_name_)
+{
+  m_serv = new ::eCAL::CServiceServer(StringToStlString(server_name_));
```
[the remark this answers is not in the document]
Comment (KerstinKeller) [reply lang/csharp/Continental/eCAL/Core/ecal_clr.cpp:413]: Sorry, but this doesn't work 😞 
```
C++ a member of a managed class cannot be of a non-managed class type
```

**Comment by lukaswerner** [comment]
LGTM
https://github.com/eclipse-ecal/ecal/pull/908 can be closed if this is merged

**Comment by KerstinKeller** [comment]
All commits from this PR to be merged by separate PRs.

**Pull request closed without merging**